File: laba4_4.cpp
#include <iostream>
#include <vector>
#include <string>
#include <map>
#include <algorithm>

using namespace std;

// Класс для представления кандидата
class Candidate {
public:
    string name;
    int borda_score;
    
    Candidate(const string& n) : name(n), borda_score(0) {}
};

// Класс для представления избирателя
class Voter {
public:
    vector<string> preferences;
    
    Voter(const vector<string>& pref) : preferences(pref) {}
};

// Класс для проведения выборов
class Election {
private:
    vector<Candidate> candidates;
    vector<Voter> voters;
    
public:
    // Добавление кандидата
    void addCandidate(const string& name) {
        candidates.emplace_back(name);
    }
    
    // Добавление избирателя
    void addVoter(const vector<string>& preferences) {
        voters.emplace_back(preferences);
    }
    
    // Метод Борда
    string calculateBordaWinner() {
        // Сброс счетчиков
        for (auto& c : candidates) {
            c.borda_score = 0;
        }
        
        // Начисление баллов
        for (const auto& voter : voters) {
            for (size_t i = 0; i < voter.preferences.size(); ++i) {
                int points = voter.preferences.size() - i - 1;
                for (auto& c : candidates) {
                    if (c.name == voter.preferences[i]) {
                        c.borda_score += points;
                        break;
                    }
                }
            }
        }
        
        // Поиск победителя
        int max_score = -1;
        string winner;
        for (const auto& c : candidates) {
            if (c.borda_score > max_score) {
                max_score = c.borda_score;
                winner = c.name;
            }
        }
        
        return winner;
    }
    
    // Метод Кондорсе
    string calculateCondorcetWinner() {
        map<string, map<string, int>> pairwise_results;
        
        // Инициализация матрицы попарных сравнений
        for (const auto& c1 : candidates) {
            for (const auto& c2 : candidates) {
                if (c1.name != c2.name) {
                    pairwise_results[c1.name][c2.name] = 0;
                }
            }
        }
        
        // Заполнение матрицы попарных сравнений
        for (const auto& voter : voters) {
            for (size_t i = 0; i < voter.preferences.size(); ++i) {
                for (size_t j = i + 1; j < voter.preferences.size(); ++j) {
                    string preferred = voter.preferences[i];
                    string other = voter.preferences[j];
                    pairwise_results[preferred][other]++;
                }
            }
        }
        
        // Поиск победителя Кондорсе
        for (const auto& c1 : candidates) {
            bool is_condorcet_winner = true;
            
            for (const auto& c2 : candidates) {
                if (c1.name != c2.name) {
                    if (pairwise_results[c1.name][c2.name] <= pairwise_results[c2.name][c1.name]) {
                        is_condorcet_winner = false;
                        break;
                    }
                }
            }
            
            if (is_condorcet_winner) {
                return c1.name;
            }
        }
        
        return "Нет победителя по Кондорсе";
    }
};

int main() {
    Election election;
    
    // Пример 1: Простой случай с явным победителем
    {
        election.addCandidate("Петя");
        election.addCandidate("Вася");
        election.addCandidate("Оля");
        
        // 2 избирателя предпочитают: Вася > Оля > Петя
        election.addVoter({"Вася", "Оля", "Петя"});
        election.addVoter({"Вася", "Оля", "Петя"});
        
        // 1 избиратель предпочитает: Оля > Вася > Петя
        election.addVoter({"Оля", "Вася", "Петя"});
        
        string borda_winner = election.calculateBordaWinner();
        string condorcet_winner = election.calculateCondorcetWinner();
        
        cout << "Пример 1:\n";
        cout << "Победитель по Борда: " << borda_winner << endl;
        cout << "Победитель по Кондорсе: " << condorcet_winner << endl;
        
        if (borda_winner != condorcet_winner) {
            cout << "Замечание: разные методы дали разных победителей!\n";
        }
        cout << "------------------------\n";
    }
    
    // Пример 2: Парадокс Кондорсе (разные победители)
    {
        Election paradox_election;
        paradox_election.addCandidate("A");
        paradox_election.addCandidate("B");
        paradox_election.addCandidate("C");
        
        // Голоса, создающие парадокс Кондорсе
        paradox_election.addVoter({"A", "B", "C"});
        paradox_election.addVoter({"B", "C", "A"});
        paradox_election.addVoter({"C", "A", "B"});
        
        string borda_winner = paradox_election.calculateBordaWinner();
        string condorcet_winner = paradox_election.calculateCondorcetWinner();
        
        cout << "Пример 2 (Парадокс Кондорсе):\n";
        cout << "Победитель по Борда: " << borda_winner << endl;
        cout << "Победитель по Кондорсе: " << condorcet_winner << endl;
        
        if (borda_winner != condorcet_winner) {
            cout << "Замечание: разные методы дали разных победителей!\n";
            cout << "Это демонстрирует парадокс Кондорсе, когда нет победителя по Кондорсе,\n";
            cout << "но метод Борда дает определенного победителя.\n";
        }
        cout << "------------------------\n";
    }
    
    // Пример 3: Все избиратели имеют одинаковые предпочтения
    {
        Election unanimous_election;
        unanimous_election.addCandidate("X");
        unanimous_election.addCandidate("Y");
        unanimous_election.addCandidate("Z");
        
        // Все избиратели предпочитают: X > Y > Z
        for (int i = 0; i < 5; ++i) {
            unanimous_election.addVoter({"X", "Y", "Z"});
        }
        
        string borda_winner = unanimous_election.calculateBordaWinner();
        string condorcet_winner = unanimous_election.calculateCondorcetWinner();
        
        cout << "Пример 3 (Единогласное голосование):\n";
        cout << "Победитель по Борда: " << borda_winner << endl;
        cout << "Победитель по Кондорсе: " << condorcet_winner << endl;
        
        if (borda_winner != condorcet_winner) {
            cout << "Замечание: разные методы дали разных победителей!\n";
        } else {
            cout << "Оба метода согласны с победителем.\n";
        }
        cout << "------------------------\n";
    }
    
    return 0;
}

File: laba4_4.go
package main

import (
	"fmt"
	"strings"
)

// Candidate представляет кандидата с его баллами
type Candidate struct {
	Name       string
	BordaScore int
}

// Voter представляет избирателя с его предпочтениями
type Voter struct {
	Preferences []string
}

// Election представляет выборы с кандидатами и избирателями
type Election struct {
	Candidates []Candidate
	Voters     []Voter
}

// AddCandidate добавляет кандидата
func (e *Election) AddCandidate(name string) {
	e.Candidates = append(e.Candidates, Candidate{Name: name})
}

// AddVoter добавляет избирателя
func (e *Election) AddVoter(preferences []string) {
	e.Voters = append(e.Voters, Voter{Preferences: preferences})
}

// CalculateBordaWinner вычисляет победителя по методу Борда
func (e *Election) CalculateBordaWinner() string {
	// Сбрасываем баллы
	for i := range e.Candidates {
		e.Candidates[i].BordaScore = 0
	}

	// Начисляем баллы
	for _, voter := range e.Voters {
		for pos, name := range voter.Preferences {
			for i := range e.Candidates {
				if e.Candidates[i].Name == name {
					e.Candidates[i].BordaScore += len(voter.Preferences) - pos - 1
					break
				}
			}
		}
	}

	// Находим победителя
	maxScore := -1
	winner := ""
	for _, c := range e.Candidates {
		if c.BordaScore > maxScore {
			maxScore = c.BordaScore
			winner = c.Name
		}
	}

	return winner
}

// CalculateCondorcetWinner вычисляет победителя по методу Кондорсе
func (e *Election) CalculateCondorcetWinner() string {
	// Создаем матрицу попарных сравнений
	pairwise := make(map[string]map[string]int)
	for _, c1 := range e.Candidates {
		pairwise[c1.Name] = make(map[string]int)
		for _, c2 := range e.Candidates {
			if c1.Name != c2.Name {
				pairwise[c1.Name][c2.Name] = 0
			}
		}
	}

	// Заполняем матрицу
	for _, voter := range e.Voters {
		for i, c1 := range voter.Preferences {
			for j := i + 1; j < len(voter.Preferences); j++ {
				c2 := voter.Preferences[j]
				pairwise[c1][c2]++
			}
		}
	}

	// Проверяем каждого кандидата на победу по Кондорсе
	for _, c1 := range e.Candidates {
		isCondorcetWinner := true
		for _, c2 := range e.Candidates {
			if c1.Name != c2.Name {
				if pairwise[c1.Name][c2.Name] <= pairwise[c2.Name][c1.Name] {
					isCondorcetWinner = false
					break
				}
			}
		}
		if isCondorcetWinner {
			return c1.Name
		}
	}

	return "Нет победителя по Кондорсе"
}

func main() {
	// Пример 1: Простой случай
	election1 := Election{}
	election1.AddCandidate("Петя")
	election1.AddCandidate("Вася")
	election1.AddCandidate("Оля")

	election1.AddVoter([]string{"Вася", "Оля", "Петя"})
	election1.AddVoter([]string{"Вася", "Оля", "Петя"})
	election1.AddVoter([]string{"Оля", "Вася", "Петя"})

	borda1 := election1.CalculateBordaWinner()
	condorcet1 := election1.CalculateCondorcetWinner()

	fmt.Println("Пример 1:")
	fmt.Println("Победитель по Борда:", borda1)
	fmt.Println("Победитель по Кондорсе:", condorcet1)
	if borda1 != condorcet1 {
		fmt.Println("Замечание: разные методы дали разных победителей!")
	}
	fmt.Println(strings.Repeat("-", 30))

	// Пример 2: Парадокс Кондорсе
	election2 := Election{}
	election2.AddCandidate("A")
	election2.AddCandidate("B")
	election2.AddCandidate("C")

	election2.AddVoter([]string{"A", "B", "C"})
	election2.AddVoter([]string{"B", "C", "A"})
	election2.AddVoter([]string{"C", "A", "B"})

	borda2 := election2.CalculateBordaWinner()
	condorcet2 := election2.CalculateCondorcetWinner()

	fmt.Println("Пример 2 (Парадокс Кондорсе):")
	fmt.Println("Победитель по Борда:", borda2)
	fmt.Println("Победитель по Кондорсе:", condorcet2)
	if borda2 != condorcet2 {
		fmt.Println("Замечание: разные методы дали разных победителей!")
		fmt.Println("Это демонстрирует парадокс Кондорсе, когда нет победителя по Кондорсе,")
		fmt.Println("но метод Борда дает определенного победителя.")
	}
	fmt.Println(strings.Repeat("-", 30))

	// Пример 3: Единогласное голосование
	election3 := Election{}
	election3.AddCandidate("X")
	election3.AddCandidate("Y")
	election3.AddCandidate("Z")

	for i := 0; i < 5; i++ {
		election3.AddVoter([]string{"X", "Y", "Z"})
	}

	borda3 := election3.CalculateBordaWinner()
	condorcet3 := election3.CalculateCondorcetWinner()

	fmt.Println("Пример 3 (Единогласное голосование):")
	fmt.Println("Победитель по Борда:", borda3)
	fmt.Println("Победитель по Кондорсе:", condorcet3)
	if borda3 != condorcet3 {
		fmt.Println("Замечание: разные методы дали разных победителей!")
	} else {
		fmt.Println("Оба метода согласны с победителем.")
	}
	fmt.Println(strings.Repeat("-", 30))
}
